modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsHolder.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cache.common.tier;
+
+import org.opensearch.common.cache.stats.DefaultCacheStatsHolder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * A tier-aware version of DefaultCacheStatsHolder. Overrides the incrementer functions, as we cannot just add the on-heap
+ * and disk stats to get a total for the cache as a whole. For example, if the heap tier has 5 misses and the disk tier
+ * has 4, the total cache has had 4 misses, not 9. The same goes for evictions. Other stats values add normally.
+ * This means for misses and evictions, if we are incrementing for the on-heap tier and the disk tier is present,
+ * we have to increment only the leaf nodes corresponding to the on-heap tier itself, and not its ancestors.
+ * If the disk tier is not present, we do increment the ancestor nodes.
```

Comment:
Could the tiered cache just expose the stats for the underlying tiers? 

It should be clear that every hit from the disk cache occurs as a result of a miss from the memory cache, a miss on the disk cache follows a miss from the memory cache, etc.

Creating a whole separate subclass of the stats holder feels unnecessarily complicated and IMO is a clear sign that we're doing it wrong.

Reply:
That wouldn't work, because of the dynamic disk tier (and also the computeIfAbsent issue Sagar mentioned). Say the disk tier is disabled midway through use, and we ask the heap tier for its misses. How would the heap tier, or the TSC, know which misses happened when the disk tier was on (shouldn't contribute to the total) vs. when the disk tier was off (should contribute to the total)? 

Also, in order to pass this to ImmutableCacheStatsHolder in a way that aggregation by levels is correct, we would have to use the two stats trees from each tier to reconstruct the same tree structure we are currently building in TieredSpilloverCacheStatsHolder anyway.

Reply:
> Also, in order to pass this to ImmutableCacheStatsHolder in a way that aggregation by levels is correct, we would have to use the two stats trees from each tier to reconstruct the same tree structure we are currently building in TieredSpilloverCacheStatsHolder anyway. Since there would be no way to do this with the existing logic in DefaultCacheStatsHolder I think we would still have to make some separate class to make this temporary tree regardless.

Each cache can hold its own `DefaultCacheStatsHoider`. The tiered cache can request immutable snapshots from each tier (even accounting for tiering being dynamic), passing in the levels. (I still think the level nonsense is way overcomplicating this stuff and should have been left for a future release, if ever). Then group the two under a top-level "tier" dimension.

Reply:
The basic problem is there's no way to accurately combine snapshots from the heap and disk tier to get total stats, either for the tiered cache as a whole, or for an individual index or shard. There's additional logic in the tiered cache itself which changes whether a hit/miss/eviction counts towards the total, and the tiers aren't and shouldn't be aware of this. 

First is the getOrCompute() issue that Sagar mentioned: 

> This is because computeIfAbsent() does a get() first and then call underlying onHeap.computeIfAbsent() which again does a get() inside to check whether value is present or not. This is needed to handle multiple concurrent requests for same key, where the first get() will return null and then only of them will load the value in onHeap.computeIfAbsent() and rest will just wait and fetch the value(resulting in hits for them).

This problem can't be sidestepped even by adding some `captureStats` parameter to ICache.get(), and passing this in different places in getOrCompute(). Imagine some number of concurrent requests ask for the same key. One will ultimately load it into the heap tier. But first we run get() in all of them and see if it's present. For each request, this initial get() should add a miss in the disk tier if this thread is the one to load the value - because we computed it from scratch. But it shouldn't add a disk miss if another thread loads it in - because in this case the value will ultimately be loaded from the heap tier. But when running this first get(), we don't know whether this thread will be the one to load it. 

Second is the dynamic disk cache issue that I mentioned in the previous comments. 

We could sidestep this by keeping track of stats for the tiered cache as well. Then when someone calls stats() we can either send the total stats (if `tier` isn't in levels) or combine the two tier trees into one (if it is). But we can't just keep track of a running total for just the tiered cache itself, we also have to do this for every index and shard in order to be able to break down stats by indices or shards, which we agreed at the beginning of the API design we wanted to do. At this point we would be storing a third DefaultCacheStatsHolder for tiered-cache total stats, instead of just one as we have in the current implementation.

Reply:
@msfroh Don't we still want to be able to return the total stats for the request cache, even if the tiered cache is used? We'd decided in the API design doc to always return the total stats, followed by any levels if present. This mirrors the existing indices request cache API.

Reply:
Even if we didn't display total stats, which I think the customer definitely needs to be able to access, this doesn't solve the getOrCompute problem. If the user makes one request to the request cache as a whole, they should see at most 1 hit/miss for each tier. But with the tiered spillover cache's getOrCompute logic, they could see more than one, which doesn't really make sense. And if they can't see the total stats show the cache as a whole only had 1 hit/miss, they don't have any way of reconciling why this is happening. 

@sgup432 @jainankitk What are your thoughts?